### PR TITLE
Separate data per P2P port

### PR DESCRIPTION
**Pull request by IPGlider**
poolstate.bin and p2pstate.bin are stored in .bitmonero/ if the default P2P port is being used.

If another port is used both files are stored in .bitmonero/PORTNUMBER/.

### CODE REVIEW

File: src/cryptonote_core/cryptonote_core.cpp [hunk 1]
```diff
@@ -259,8 +260,17 @@ namespace cryptonote
...
     m_fakechain = test_options != NULL;
     bool r = handle_command_line(vm);
+    bool testnet = command_line::get_arg(vm, command_line::arg_testnet_on);
+    auto p2p_bind_arg = testnet ? nodetool::arg_testnet_p2p_bind_port : nodetool::arg_p2p_bind_port;
+    std::string m_port = command_line::get_arg(vm, p2p_bind_arg);
+    std::string m_config_folder_mempool = m_config_folder;
+
+    if ((!testnet && m_port != std::to_string(::config::P2P_DEFAULT_PORT))
+        || (testnet && m_port != std::to_string(::config::testnet::P2P_DEFAULT_PORT))) {
+      m_config_folder_mempool = m_config_folder_mempool + "/" + m_port;
```
Comment (moneromooo-monero) [inline src/cryptonote_core/cryptonote_core.cpp]: Can this break on Windows ? The rest of the code uses boost::path, which has a / operator for this.

File: src/p2p/net_node.inl [hunk 2]
```diff
@@ -73,16 +73,6 @@ namespace nodetool
...
-    const command_line::arg_descriptor<std::string> arg_p2p_bind_port = {
-        "p2p-bind-port"
-      , "Port for p2p network protocol"
-      , std::to_string(config::P2P_DEFAULT_PORT)
-      };
-    const command_line::arg_descriptor<std::string> arg_testnet_p2p_bind_port = {
-        "testnet-p2p-bind-port"
-      , "Port for testnet p2p network protocol"
-      , std::to_string(config::testnet::P2P_DEFAULT_PORT)
-      };
```
Comment (moneromooo-monero) [inline src/p2p/net_node.inl]: Why did you need to move this ? net_node.h includes net_node.inl., and I'd rather keep the whole option set in the same place.

File: src/p2p/net_node.inl [hunk 3]
```diff
@@ -505,6 +495,11 @@ namespace nodetool
     auto config_arg = testnet ? command_line::arg_testnet_data_dir : command_line::arg_data_dir;
     m_config_folder = command_line::get_arg(vm, config_arg);
 
+    if ((!testnet && m_port != std::to_string(::config::P2P_DEFAULT_PORT))
+        || (testnet && m_port != std::to_string(::config::testnet::P2P_DEFAULT_PORT))) {
+      m_config_folder = m_config_folder + "/" + m_port;
```
Comment (moneromooo-monero) [inline src/p2p/net_node.inl]: Same as above about /

File: src/cryptonote_core/cryptonote_core.cpp (hunk 1 above)
Comment (IPGlider) [inline src/cryptonote_core/cryptonote_core.cpp]: Haven't tested on Windows. I took inspiration from https://github.com/monero-project/monero/blob/master/src/p2p/net_node.inl#L667.

Will take a look at boost::filesystem::path.

File: src/p2p/net_node.inl (hunk 2 above)
Comment (IPGlider) [inline src/p2p/net_node.inl]: Moved it to the header file to be able to include it on cryptonote_core.cpp and use the values in the if statement.

Do you recommend me another way to do it?

Comment (IPGlider) [inline src/p2p/net_node.inl]: Just got what you meant. Will change it.

File: src/cryptonote_core/cryptonote_core.cpp (hunk 1 above)
Comment (IPGlider) [inline src/cryptonote_core/cryptonote_core.cpp]: On other parts of the system boost::filesystem::path is used as you denote, but in those two cases std::string is used.

In the link from my previous comment and in the following one slashes are already used, so behaviour should not change.

https://github.com/monero-project/monero/blob/master/src/cryptonote_core/tx_pool.cpp#L714

**Comment by hyc** [comment]
This PR appears to be broken, to me. Nobody creates the subdir for the state files. Perhaps we should just extend the filename instead of using a subdir.
